### ISSUE TITLE
Security: allow restrict custom HTTP request header to specific URL patterns

### ISSUE THREAD

**Issue by sanmai-NL** [lycheeverse/lychee#1298]
Currently, the `--header` CLI and configuration parameter does not distinguish between target URLs. If the user where to supply a crucial secret in there, in order to make some URLs accessible during CI, then this secret would leak to all other hosts for which URLs are found.

## Solution design

Rather than implementing URL/header mapping logic in Lychee, I propose to separate this concern into an, if you so choose, external tool like a proxy. If Lychee were to have proxying support, complex logic, mappings, analysis, flows, etc. can be configured through a proxy. If common use cases are documented in a how-to style within the Lychee docs, the value for the user would not be much less than with a native implementation by Lychee (one could argue, much more indeed).

**Comment by sanmai-NL** [comment]
An example of a tool that could support this functionality: https://docs.mitmproxy.org/stable/overview-features/#modify-headers

**Comment by mre** [comment]
I like the `modify_headers` syntax you linked to. We could add something like this.
Just to clarify, this doesn't require a proxy, but rather a way to pass these headers to `reqwest`, our HTTP request client, right?

**Comment by thomas-zahner** [comment]
I definitely agree that there should be the possibility to scope headers to specific hosts. The mentioned syntax looks really interesting. But I didn't find a proper specification for that syntax. It might also be a bit overkill to start with that syntax. For the beginning it would suffice if we can somehow associate host names with header values. We could still use that syntax but only implement the first example.

> Just to clarify, this doesn't require a proxy, but rather a way to pass these headers to reqwest, our HTTP request client, right?

@mre I also understand it that way.

**Comment by thomas-zahner** [comment]
Also for the person implementing this: try to make use of [the internal request chain](https://docs.rs/lychee-lib/latest/lychee_lib/chain/type.RequestChain.html) if possible.

**Comment by thomas-zahner** [comment]
https://github.com/lycheeverse/lychee/pull/1929 should address this issue with a different approach. [It clarifies](https://github.com/lycheeverse/lychee/pull/1929/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R527) that `--header` is used for ALL requests. The PR introduces [`HostConfig`](https://github.com/lycheeverse/lychee/pull/1844/files#diff-fb597381f4e45b19e12ebefa29ba05bf616b01f6c38055c5e57ea558fd2e7701R44) which includes headers. These headers are only sent to specific hosts, i.e. specific domains/subdomains.

This approach is really clean IMO because we do not need any new complicated syntax. Instead we make good use of the TOML syntax in the config file.

```toml
[hosts."localhost"]
headers = { "A" = "B" } # these headers are only sent to localhost
```

**Comment by sanmai-NL** [comment]
> https://github.com/lycheeverse/lychee/pull/1929 should address this issue with a different approach. [It clarifies](https://github.com/lycheeverse/lychee/pull/1929/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R527) that `--header` is used for ALL requests. The PR introduces [`HostConfig`](https://github.com/lycheeverse/lychee/pull/1844/files#diff-fb597381f4e45b19e12ebefa29ba05bf616b01f6c38055c5e57ea558fd2e7701R44) which includes headers. These headers are only sent to specific hosts, i.e. specific domains/subdomains.
> 
> This approach is really clean IMO because we do not need any new complicated syntax. Instead we make good use of the TOML syntax in the config file.
> 
> ```toml
> [hosts."localhost"]
> headers = { "A" = "B" } # these headers are only sent to localhost
> ```

Thanks. How does this address the fact that some if not most of these headers may be secret (and e.g., in a secrets path mounted as filesystem)?

**Comment by thomas-zahner** [comment]
That's a great question. I have currently three ideas in mind. What do you think @sanmai-NL and @mre?

## Secret toml

Adding something like `--secrets secret.toml` where the values of `secret.toml` would be merged into the main config. This seems to align with how [Shuttle handles secrets](https://docs.shuttle.dev/resources/shuttle-secrets). I'm not sure how difficult it is to merge TOML values. (would probably use something like [figment](https://docs.rs/figment/latest/figment/) or [config](https://crates.io/crates/config/)) 

## Ask when encountering sentinel value

Users could specify something like `headers = { "A" = 0 }` or `headers = { "A" = "<INTERACTIVE>" }`. When we encounter such a value lychee prompts for the value at runtime. (e.g. like sudo)

After thinking about it this approach feels like a hack and is actually very limiting, because there is no way to provide the values programmatically.

## Support environment variables

Read environment variables, e.g. `headers = { "C" = "${SECRET_HEADER_C}" }`. We could then make lychee fail or even prompt the user as in the above idea if the environment variable is unset. This seems like the simplest feasible approach to me, with the downside that it might not be very idiomatic and common in TOML.

**Comment by sanmai-NL** [comment]
The first one is most robust and secure, and doesn't exclude e.g. the last one.

**Comment by thomas-zahner** [comment]
True, the three approaches aren't mutually exclusive. So we could prioritise them in that order: secret TOML and environment variables. (and then maybe the sentinel value for an interactive prompt)

**Comment by thomas-zahner** [comment]
@sanmai-NL It's now possible to specify multiple config files, which are merged. I figured it's much cleaner and obvious to just allow multiple `--config` flags instead of creating a dedicated `--secrets` flag. So you could now do `lychee --config lychee.toml --config secrets.toml`.
